integration/spark/src/main/spark3/java/io/openlineage/spark3/agent/lifecycle/plan/DataSourceV2RelationVisitor.java

```diff
@@ -106,16 +127,52 @@ private Provider findDatasetProvider(LogicalPlan plan) {
             .build());
   }
 
+  private OpenLineage.Dataset findDatasetForAzureCosmos(DataSourceV2Relation relation) {
+    String namespace = "azurecosmos";
+    String relationName = relation.table().name().replace("com.azure.cosmos.spark.items.", "");
+    int expectedParts = 3;
+    String[] tableParts = relationName.split("\\.", expectedParts);
+    String tableName;
+    if (tableParts.length != expectedParts) {
+      tableName = relationName;
+    } else {
+      tableName =
+          String.format(
+              "https://%s.documents.azure.com/dbs/%s/colls/%s",
+              tableParts[0], tableParts[1], tableParts[2]);
+    }
+    return PlanUtils.getDataset(
+        tableName,
+        namespace,
+        new OpenLineage.DatasetFacetsBuilder()
+            .schema(schemaFacet(relation.schema()))
+            .dataSource(datasourceFacet(namespace))
+            .put(
+                "table_provider",
+                new TableProviderFacet(
+                    Provider.AZURECOSMOS.name().toLowerCase(), "json")) // Cosmos is always json
+            .build());
+  }
+
   @Override
   public List<OpenLineage.Dataset> apply(LogicalPlan logicalPlan) {
-    Provider provider = findDatasetProvider(logicalPlan);
-    DataSourceV2Relation x = (DataSourceV2Relation) logicalPlan;
+    LogicalPlan plan = null;
+    if (isV2ScanRelation(logicalPlan)) {
+      plan = (LogicalPlan) ((DataSourceV2ScanRelation) logicalPlan).relation();
+    } else {
+      plan = logicalPlan;
+    }
+
+    Provider provider = findDatasetProvider(plan);
+    DataSourceV2Relation x = (DataSourceV2Relation) plan;
 
     switch (provider) {
       case ICEBERG:
         return nullableSingletonList(findDatasetForIceberg(x));
       case DELTA:
         return nullableSingletonList(findDatasetForDelta(x));
+      case AZURECOSMOS:
```

Comment:
Is it possible to provide any integration test for that? It would be super beneficial to make sure the feature does not get broken later on.

Reply:
I can definitely provide a more complete integration test.  I looked for an example working with some cloud infrastructure (e.g. for BigQuery) but I didn't find anything where you guys were spinning up a GCP cluster and firing off queries there.  [Cosmos Db does have a free tier](https://docs.microsoft.com/en-us/azure/cosmos-db/free-tier) which I could create / use for the integration test.

Is that what you're looking for pawel?

Reply:
From my perspective, at some point we'd want to do that (as we do with Airflow integration and GCP https://github.com/OpenLineage/OpenLineage/blob/main/.circleci/config.yml#L309) but it's not required for this PR.